rust/hook-api/src/handlers/webhook.rs

```diff
@@ -95,6 +95,19 @@ pub async fn post_hoghook(
 ) -> Result<Json<WebhookPostResponse>, (StatusCode, Json<WebhookPostResponse>)> {
     debug!("received payload: {:?}", payload);
 
+    // We use these fields for metrics in the janitor, but we don't actually need to do anything
+    // with them now.
+    payload
```

Comment:
If/when we start looking at a v2 of this, I'd rather have required fields at the type level, to let serde do this kind of checking for us - but it's fine for now.

Reply:
For context: the reason we operate on a `Value` is because we need to retain the entire original payload, and don't want to know anything about it's shape. (We use the whole original `Value` at the end of the fn to store into Postgres, so we can use it later to send to Kafka.)

I'd love to use a struct here, but I couldn't find a way to do it without cloning the entire payload? Maybe you know a way?

```rust
#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
struct MinimalMetadata {
    team_id: u32,
    hog_function_id: String,
}

// Can't use a reference here, and can't clone just to part I want, which is why I manually inspect the 2 fields for now.
let _: MinimalMetadata = serde_json::from_value(&payload).map_err(|err| {
    bad_request(format!("unable to deserialize minimal metadata: {}", err))
})?;
```

Reply:
My first instinct is to jump to something like[ `#[serde(flatten)]`](https://serde.rs/attr-flatten.html#capture-additional-fields), but I haven't sat down to write the specific struct definition to use it yet.

Reply:
Something like this should work I think (and simplifies that function a bit):
```rust
#[derive(Debug, Serialize, Deserialize)]
struct HogPayload {
    #[serde(rename = "teamId")]
    team_id: i64,
    #[serde(rename = "hogFunctionId")]
    hog_function_id: String,
    #[serde(rename = "asyncFunctionRequest")]
    async_function_request: HoghookAsyncFunctionRequest,
    #[serde(flatten)]
    other: HashMap<String, Value>
}
```

Reply:
I tried that before, if we serialized that into `metadata` in Postgres then it would alter the original payload (putting other fields into `other`) right? Or how are you using it?

I'm surprised tests pass? I think we need better tests here. :)

Reply:
Ah, I was roundtripping back to a JSON value before writing to PG - for my hacky test it was just this on line 167:

```rust
let payload = serde_json::to_value(payload).map_err(|e| bad_request(format!("invalid json: {}", e)))?;
let job = NewJob::new(max_attempts, payload, parameters, url_hostname.as_str());
```

The way `flatten` is supposed to work is that anything in `other` at serialisation time get's pushed back out into the top level of the json (no `other` field should exist in the output object). I don't know how sqlx handles flatten fields otherwise.

Reply:
We discussed in sync, but I definitely derped and didn't imagine it would serialize it back in place. I'll switch to this in the future if we keep this code.